File: discord-bot/discord_bot/crypto.py
"""AES-256-GCM encryption helpers for storing API tokens in discord_configs.

The encrypted_token column stores a base64-encoded blob of:
    nonce (12 bytes) || ciphertext+tag

The plaintext is a JSON payload: {"token": "as_...", "site_name": "..."}
Bundling both values avoids needing an extra Supabase lookup on every message.
"""

import base64
import json
import os

from cryptography.hazmat.primitives.ciphers.aead import AESGCM


def _key() -> bytes:
    from discord_bot.config import DISCORD_AES_KEY
    return base64.b64decode(DISCORD_AES_KEY)


def encrypt_payload(token: str, site_name: str) -> str:
    """Encrypt token + site_name as AES-256-GCM. Returns base64(nonce||ciphertext+tag)."""
    plaintext = json.dumps({"token": token, "site_name": site_name}).encode()
    nonce = os.urandom(12)
    ct = AESGCM(_key()).encrypt(nonce, plaintext, None)
    return base64.b64encode(nonce + ct).decode()


def decrypt_payload(encrypted: str) -> tuple[str, str]:
    """Decrypt an encrypted_token blob. Returns (raw_token, site_name)."""
    data = base64.b64decode(encrypted)
    nonce, ct = data[:12], data[12:]
    payload = json.loads(AESGCM(_key()).decrypt(nonce, ct, None))
    return payload["token"], payload["site_name"]

File: messaging-bot/src/crypto.ts
/**
 * AES-256-GCM helpers for encrypting API tokens stored in messaging_configs.
 *
 * Wire format matches the Python discord-bot crypto.py:
 *   base64( nonce(12 bytes) || ciphertext+tag )
 *
 * Plaintext is JSON: {"token": "as_...", "site_name": "..."}
 */

import { createCipheriv, createDecipheriv, randomBytes } from 'node:crypto'
import { AES_KEY } from './config.js'

function key(): Buffer {
  return Buffer.from(AES_KEY, 'base64')
}

export function encryptPayload(token: string, siteName: string): string {
  const plaintext = Buffer.from(JSON.stringify({ token, site_name: siteName }), 'utf8')
  const nonce = randomBytes(12)
  const cipher = createCipheriv('aes-256-gcm', key(), nonce)
  const encrypted = Buffer.concat([cipher.update(plaintext), cipher.final()])
  const tag = cipher.getAuthTag()
  return Buffer.concat([nonce, encrypted, tag]).toString('base64')
}

export function decryptPayload(encrypted: string): { token: string; siteName: string } {
  const data = Buffer.from(encrypted, 'base64')
  const nonce = data.subarray(0, 12)
  const tag = data.subarray(data.length - 16)
  const ciphertext = data.subarray(12, data.length - 16)
  const decipher = createDecipheriv('aes-256-gcm', key(), nonce)
  decipher.setAuthTag(tag)
  const plaintext = Buffer.concat([decipher.update(ciphertext), decipher.final()])
  const payload = JSON.parse(plaintext.toString('utf8'))
  return { token: payload.token, siteName: payload.site_name }
}
